### DISCUSSION TITLE
`unsafe_raw` with yield renders the yielded content outside

### DISCUSSION THREAD

**Discussion by davidalejandroaguilar** [yippee-fun/phlex#793] [Help] [answered]
Most likely this is not an issue, but rather how it's intended to work, so I'm opening a discussion instead of an issue to see if someone knows an alternative way to do this:

```rb
class MyComponent < Phlex::HTML
  def view_template
    unsafe_raw "<div>#{yield}</div>"
  end
end
```

```rb
render MyComponent.new do
  "Hello"
end
```

Expected:

```html
<div>Hello</div>
```

Got:

```html
Hello<div></div>
```

Closest discussion is [Discussion #368: Block rendering outside of component template?](https://github.com/orgs/phlex-ruby/discussions/368), but that's for `ViewComponent`. This is only using Phlex.

**Comment by joeldrapper** [comment, marked as answer]
That’s because this
```ruby
unsafe_raw "<div>#{yield}</div>"
```
Is the same as this
```ruby
output = yield
unsafe_raw "<div>#{output}</div>"
```

But if the block is doing Phlex HTML stuff, it’s not going to return anything it’s just going to push stuff to the buffer. Instead of yielding, you could capture the block to a string.

```ruby
unsafe_raw "<div>#{capture(&)}</div>"
```

`capture` will temporarily change the internal buffer to a new string, yield, put the old buffer back and return the string.

**Comment by davidalejandroaguilar** [comment]
Ahh, there we go, thanks a lot!

> it’s not going to return anything it’s just going to push stuff to the buffer

> `capture` will temporarily change the internal buffer to a new string, yield, put the old buffer back and return the string.

And also thank you for the additional information, very useful.
